### PR TITLE
Write the shard generator for the new reindexer

### PR DESCRIPTION
**Pull request by alexwlchan**
When we get rows out of the table, look to see if they a default (or missing!) reindexShard. If so, create a new reindex shard and re-put to the table.

### CODE REVIEW

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    dynamodb_client = dynamodb_client or boto3.client('dynamodb')
+    table_name = os.environ['TABLE_NAME']
+
+    for sns_event in extract_sns_messages_from_lambda_event(event):
+        row = sns_event.message
+
+        # If we've already looked at this row before, there's nothing
+        # for us to do -- we can skip to the next row.
+        if row.get('reindexShard') != 'default':
```
Comment (kenoir) [inline catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py]: It still might be worth checking that the shard we generate for this ID matches the one given?

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        current_row = dynamodb_client.get_item(
+            TableName=table_name,
+            Key={'id': {'S': row['id']}},
+            AttributesToGet=['version']
+        )
+        current_version = current_row['Item']['version']['N']
+
+        # Then we call UpdateItem.  We only need to change the version and
+        # the reindexShard fields, and we condition it on the version
+        # incrementing by 1.
```
Comment (kenoir) [inline catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py]: 👍

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py (hunk 1 above)
Comment (alexwlchan) [reply catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py]: Good point; I’ll change it.

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    for sns_event in extract_sns_messages_from_lambda_event(event):
+        row = sns_event.message
+
+        new_reindex_shard = create_reindex_shard(
+            source_id=row['sourceId'],
+            source_name=row['sourceName'],
+            source_size=SOURCE_SIZES[row['sourceName']]
+        )
+
+        # If the reindex shard doesn't match the current schema, we'll
```
Comment (kenoir) [inline catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py]: Does this comment match the code?

**Comment by alexwlchan** [comment]
We want to think about the sharding scheme a bit more before we merge this.

File: catalogue_pipeline/Makefile [hunk 4]
```diff
@@ -5,11 +5,16 @@ STACK_ROOT 	= catalogue_pipeline
...
+LAMBDAS 	= reindexer_v2/complete_reindex \
+              reindexer_v2/reindex_job_creator \
+              reindexer_v2/reindex_shard_generator
 
 TF_NAME 	= catalogue_pipeline
 TF_PATH 	= $(ROOT)/catalogue_pipeline/terraform
 
 TF_IS_PUBLIC_FACING = false
 
+reindexer_lambdas-test: reindexer_v2/reindex_job_creator-test \
```
Comment (alexwlchan) [inline catalogue_pipeline/Makefile]: I think we can delete this Make task; it was meant to be used in Travis but in practice we're just doing separate tests for each Lambda.

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                UpdateExpression='SET version = :newVersion, reindexShard=:reindexShard',
+                ConditionExpression='version < :newVersion',
+                ExpressionAttributeValues={
+                    ':newVersion': {'N': str(version + 1)},
+                    ':reindexShard': {'S': new_reindex_shard},
+                }
+            )
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] != 'ConditionalCheckFailedException':
+                raise
```
Comment (alexwlchan) [inline catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/reindex_shard_generator.py]: Worth a log if it we do have a failed condition?

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/shard_manager.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+The length of the shard varies based on the data source, to keep individual
+reindex shards for each source at a manageable size.
+
+"""
+
+import math
+
+
+def create_reindex_shard(source_name, source_id, source_size, shard_size):
+    ascii_source_id = int(''.join(str(ord(c)) for c in source_id))
```
Comment (alexwlchan) [inline catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/shard_manager.py]: Neat. I’d been thinking of something slightly different – `sum(ord(c) …)` – but I’m not sure it makes a huge amount of difference.

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/test_reindex_shard_generator.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    #   of CREATING . After the table is created, DynamoDB sets the
+    #   TableStatus to ACTIVE . You can perform read and write operations
+    #   only on an ACTIVE table.
+    #
+    #   You can use the DescribeTable action to check the table status.
+    #
+    # So we wait for the table to be created before yielding to the test.
+    #
+    # Link:  https://boto3.readthedocs.io/en/stable/reference/services/dynamodb.html#DynamoDB.Client.create_table
+    for i in range(10):
```
Comment (alexwlchan) [inline catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/test_reindex_shard_generator.py]: nit (aimed at me, I think I wrote this code originally): we could write `for _ in range(10)`, as we never use the variable `i` in the loop.

File: catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/test_reindex_shard_generator.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    main(event=event, dynamodb_client=dynamodb_client)
+
+    item = dynamodb_client.get_item(
+        TableName=source_data_table,
+        Key={'id': {'S': 'sierra/b3333333'}}
+    )
+    assert item['Item']['version']['N'] == '3'
+
+
+def test_does_not_override_record_with_old_version(dynamodb_client, source_data_table):
```
Comment (alexwlchan) [inline catalogue_pipeline/reindexer_v2/reindex_shard_generator/src/test_reindex_shard_generator.py]: 👍

File: catalogue_pipeline/terraform/iam_role_policy.tf [hunk 9]
```diff
@@ -38,3 +38,10 @@ resource "aws_iam_role_policy" "allow_reindex_job_creator_publish_to_sns" {
   role   = "${module.reindex_job_creator_lambda.role_name}"
   policy = "${module.reindex_jobs_topic.publish_policy}"
 }
+
+# Role policies for the shard generator lambda
+
+resource "aws_iam_role_policy" "allow_shard_generator_to_access_vhs" {
+  role   = "${module.shard_generator_lambda.role_name}"
+  policy = "${module.versioned-hybrid-store.full_access_policy}"
```
Comment (alexwlchan) [inline catalogue_pipeline/terraform/iam_role_policy.tf]: Does it need full access (in particular, does it need to do Reads?). Would PutItem be enough?

Comment (alicefuzier) [reply catalogue_pipeline/terraform/iam_role_policy.tf]: 👍

**Pull request merged**